src/common/Utilities.cpp

```diff
@@ -0,0 +1,133 @@
+/*
+ * Copyright 2016 The Android Open Source Project
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+
+#include <unistd.h>
+#include "oboe/Definitions.h"
+#include "oboe/Utilities.h"
+
+#define OBOE_CASE_ENUM(name) case name: return #name
```

Comment:
I tried this and it just outputs the type name e.g. "Result::OK" twice in a row. I think I need to static_cast `name` to an `int32_t` then concat with the return string - could be tough to implement in a macro.

Reply:
Right, for Kevin's approach to work, "name" must be a preprocessor define, so it can be expanded into a value during macro expansion phase, like described here: https://gcc.gnu.org/onlinedocs/gcc-4.1.2/cpp/Stringification.html. Preprocessor knows nothing about the values of enums.

But actually, the tricky part that involves the use of preprocessor is getting the program symbol name, that is, the name of the enum constant. The value of a enum constant can be obtained at compile- or run-time.

So, for example, in your enum to string function you could use a stringstream, where you first put the value of the enum (unconditionally), and then use the stringification trick in order to add the symbol name of the enum.

Reply:
Right, I though the name was an AAudio constant. There are solutions to convert the enum value to a compile time string, but it would add too much complexity. Hopefully in C++20, reflection will make all that much easier.
The alternative would be to construct the string at build time which would make it much easier:

```C++
template <class Enum, Enum enum>
const char* enumToStr(const char* name) {
    const static str = name
}
#define OBOE_CASE_ENUM(name) case name: return std::string(#name) + " (" + std::to_string(name) + ')';
```

But that changes the API, as a std::string has to be returned.